### PR TITLE
tolerate two cases of imprecision in value_set_fi; addresses issue #2653

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

**Review by tautschnig — DISMISSED** [review]
Could a test please be added? Nit pick: will github recognise this commit message as one closing the issue?

File: src/pointer-analysis/value_set_fi.cpp [hunk 1]
```diff
@@ -409,13 +409,13 @@ void value_set_fit::get_value_set_rec(
...
-    DATA_INVARIANT(type.id()==ID_array ||
-                   type.id()==ID_incomplete_array ||
-                   type.id()=="#REF#",
-                   "operand 0 of index expression must be an array");
-
-    get_value_set_rec(expr.op0(), dest, "[]"+suffix,
-                      original_type, ns, recursion_set);
+    if(type.id() == ID_array)
+    {
+      get_value_set_rec(expr.op0(), dest, "[]"+suffix,
```
Comment (smowton) [inline src/pointer-analysis/value_set_fi.cpp]: Formatting nitpick: break after open-paren

File: src/pointer-analysis/value_set_fi.cpp [hunk 2]
```diff
@@ -1328,9 +1328,10 @@ void value_set_fit::assign_rec(
 
     assign_rec(typecast_expr.op(), values_rhs, suffix, ns, recursion_set);
   }
-  else if(lhs.id()=="zero_string" ||
-          lhs.id()=="is_zero_string" ||
-          lhs.id()=="zero_string_length")
+  else if(lhs.id() == "zero_string" ||
```
Comment (smowton) [inline src/pointer-analysis/value_set_fi.cpp]: same goes here

File: src/pointer-analysis/value_set_fi.cpp (hunk 1 above)
Comment (smowton) [reply src/pointer-analysis/value_set_fi.cpp]: (git-clang-format will do this afaik)

**Review by smowton — APPROVED** [review]

File: src/pointer-analysis/value_set_fi.cpp (hunk 2 above)
Comment (kroening) [reply src/pointer-analysis/value_set_fi.cpp]: both done

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: fcdd39b2a1840aa1745737d5e4870d84a9aa17d0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84594015
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 3d243543adb4ff450596e7994e2fa1d590ec1e1b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84596572

**Comment by NlightNFotis** [comment]
Hi @kroening, do we want to keep this? Or should this be closed?

**Review by chris-ryder — APPROVED** [review]

**Pull request merged**